app/src/main/java/com/zeapo/pwdstore/crypto/BasePgpActivity.kt

```diff
@@ -0,0 +1,285 @@
+
+    /**
+     * Name of the password file
+     *
+     * Converts personal/auth.foo.org/john_doe@example.org.gpg to john_doe.example.org.gpg
+     */
+    val name: String by lazy { File(fullPath).nameWithoutExtension }
+
+    /**
+     * Get the timestamp for when this file was last modified.
+     */
+    val lastChangedString: CharSequence by lazy {
+        getLastChangedString(
+            intent.getLongExtra(
+                "LAST_CHANGED_TIMESTAMP",
+                -1L
+            )
+        )
+    }
+
+    /**
+     * [SharedPreferences] instance used by subclasses to persist settings
+     */
+    val settings: SharedPreferences by lazy { PreferenceManager.getDefaultSharedPreferences(this) }
+
+    /**
+     * [ClipboardManager] instance used by [copyTextToClipboard]. Subclasses should not be using it
+     * directly, and instead calling [copyTextToClipboard] or [copyPasswordToClipboard] as required.
+     */
+    private val clipboard by lazy { getSystemService<ClipboardManager>() }
+
+    /**
+     * Backing property and public read-only field for getting the list of OpenPGP key IDs that we
+     * have access to.
+     */
+    private var _keyIDs = emptySet<String>()
+    val keyIDs get() = _keyIDs
+
+    /**
+     * Handle to the [OpenPgpApi] instance that is used by subclasses to interface with OpenKeychain.
+     */
+    private var serviceConnection: OpenPgpServiceConnection? = null
+    var api: OpenPgpApi? = null
+
+    /**
+     * [onCreate] sets the window up with the right flags to prevent auth leaks through screenshots
+     * or recent apps screen and fills in [_keyIDs] from [settings]
+     */
+    @CallSuper
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        window.setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE)
+        tag(TAG)
+
+        _keyIDs = settings.getStringSet("openpgp_key_ids_set", null) ?: emptySet()
+    }
+
+    /**
+     * [onDestroy] handles unbinding from the OpenPgp service linked with [serviceConnection]. This
+     * is annotated with [CallSuper] because it's critical to unbind the service to ensure we're not
+     * leaking things.
+     */
+    @CallSuper
+    override fun onDestroy() {
+        super.onDestroy()
+        serviceConnection?.unbindFromService()
+    }
+
+    /**
+     * Sets up [api] once the service is bound. Downstream consumers must call super this to
+     * initialize [api]
+     */
+    @CallSuper
+    override fun onBound(service: IOpenPgpService2) {
+        api = OpenPgpApi(this, service)
+    }
+
+    /**
+     * Mandatory error handling from [OpenPgpServiceConnection.OnBound]. All subclasses must handle
+     * their own errors, and hence this class simply logs and rethrows. Subclasses Must NOT call super.
+     */
+    override fun onError(e: Exception) {
+        e { "Callers must handle their own exceptions" }
```

Comment:
I think that we should either not log here (as the Exception might be expected by the caller) or also log the exception in case the caller forgets:

```suggestion
        e(e) { "Callers must handle their own exceptions" }
```

Reply:
Rethrowing the exception should also log it, no?

Reply:
Also true.